src/Avalonia.Base/Platform/Storage/FileIO/StorageBookmarkHelper.cs

```diff
@@ -0,0 +1,128 @@
+
+        var arrayLength = PrefixLength + nativeBookmarkBytes.Length;
+        var arrayPool = ArrayPool<byte>.Shared.Rent(arrayLength);
+        try
+        {
+            // Write platform into first 16 bytes.
+            var arraySpan = arrayPool.AsSpan(0, arrayLength);
+            platform.CopyTo(arraySpan);
+
+            // Write bookmark bytes.
+            nativeBookmarkBytes.CopyTo(arraySpan.Slice(PrefixLength));
+
+            // We must use span overload because ArrayPool might return way too big array. 
+#if NET6_0_OR_GREATER
+            return Convert.ToBase64String(arraySpan);
+#else
+            return Convert.ToBase64String(arraySpan.ToArray(), Base64FormattingOptions.None);
+#endif
+        }
+        finally
+        {
+            ArrayPool<byte>.Shared.Return(arrayPool);
+        }
+    }
+
+    public static bool TryDecodeBookmark(ReadOnlySpan<byte> platform, string? base64bookmark, [NotNullWhen(true)] out byte[]? nativeBookmark)
+    {
+        if (platform.Length > PrefixLength
+            || platform.Length == 0
+            || base64bookmark is null
+            || base64bookmark.Length % 4 != 0)
+        {
+            nativeBookmark = null;
+            return false;
+        }
+
+        Span<byte> decodedBookmark;
+#if NET6_0_OR_GREATER
+        // Each base64 character represents 6 bits, but to be safe, 
+        var arrayPool = ArrayPool<byte>.Shared.Rent(PrefixLength + base64bookmark.Length * 6);
+        if (Convert.TryFromBase64Chars(base64bookmark, arrayPool, out int bytesWritten))
+        {
+            decodedBookmark = arrayPool.AsSpan().Slice(0, bytesWritten);
+        }
+        else
+        {
+            nativeBookmark = null;
+            return false;
+        }
+#else
+        decodedBookmark = Convert.FromBase64String(base64bookmark).AsSpan();
+#endif
+        try
+        {
+            if (decodedBookmark.Length < PrefixLength)
+            {
+                nativeBookmark = null;
+                return false;
+            }
+
+            var actualPlatform = decodedBookmark.Slice(0, platform.Length);
+            if (actualPlatform.SequenceEqual(platform))
+            {
+                nativeBookmark = decodedBookmark.Slice(PrefixLength).ToArray();
+                return true;
+            }
+
+            nativeBookmark = null;
+            return false;
+        }
+        finally
+        {
+#if NET6_0_OR_GREATER
+            ArrayPool<byte>.Shared.Return(arrayPool);
+#endif
+        }
+    }
+
+    public static string EncodeBclBookmark(string localPath) => EncodeBookmark(FakeBclBookmarkPlatform, localPath);
+
+    public static bool TryDecodeBclBookmark(string nativeBookmark, [NotNullWhen(true)] out string? localPath)
+    {
+        if (TryDecodeBookmark(FakeBclBookmarkPlatform, nativeBookmark, out var bytes))
+        {
+            localPath = Encoding.UTF8.GetString(bytes);
+            return true;
+        }
+
+        localPath = null;
```

Comment:
We should probably also check for the bookmark format created by older versions. Depending on the platform that would be a `StartsWith("/")` /`Regex.IsMatch(@"^[A-Za-z]:[\\/]")` check. Could also follow with a `File.Exists` check to be sure.

Reply:
Supporting older bookmarks would make this change much more difficult. Primarily because of iOS and Browser platforms, where bookmarks are the most important. We can't simply check if file exist there.
BCL implementation of bookmarks is the least important.

Reply:
So, we are planning to break all apps that were previously using the API, right?

Reply:
@kekekeks pushed couple of changes.
1. All bookmarks are now prefixed with "ava."
2. All bookmarks now have bookmark version information prefixed as well, i.e. "ava.v1.bcl"
3. If bookmark doesn't have a prefix, attempt to load it as if it's an old format bookmarks (i.e. 11.0, 11.1 state)

So should be better now.

Reply:
If we will ever need to extend bookmarks format, i.e. add extra information or make header bigger or dynamic, v2 can be added without breaking v1.